sys/include/net/gcoap/dns.h

```diff
@@ -0,0 +1,269 @@
+#endif
+
+/**
+ * @brief   allow unencrypted communication
+ *
+ * @warning Should only be set to 1 for debugging and testing!
+ */
+#ifndef CONFIG_GCOAP_DNS_ALLOW_UNENCRYPTED
+#define CONFIG_GCOAP_DNS_ALLOW_UNENCRYPTED          0
+#endif
+/** @} */
+
+/**
+ * @brief   Forward type declaration of struct gcoap_dns_ctx
+ */
+typedef struct gcoap_dns_ctx gcoap_dns_ctx_t;
+
+/**
+ * @brief   Callback for asynchronous operation
+ *
+ * @note    Requires module `gcoap_dns_async` to be used.
+ *
+ * @param[in] The context for the asynchronous DNS query.
+ */
+typedef void (*gcoap_dns_async_cb_t)(gcoap_dns_ctx_t *ctx);
+
+/**
+ * @brief   Context for a DNS query-response-pair.
+ */
+struct gcoap_dns_ctx {
+    union {
+        /**
+         * @brief   Synchronization mutex for synchronous use
+         */
+        mutex_t resp_wait;
+        /**
+         * @brief   Callback for asynchronous use
+         */
+        gcoap_dns_async_cb_t cb;
+    } sync;                 /**< Synchronization primitives */
+    /**
+     * @brief The CoAP request packet
+     *
+     * Only needs to have coap_pkt_t::payload and coap_pkt_t::payload_len
+     * initialized.
+     */
+    coap_pkt_t *pkt;
+    void *dns_buf;          /**< The buffer for the DNS message exchange */
+    void *addr_out;         /**< Pointer to the resulting address */
+    /**
+     * @brief   Status for the DNS message exchange
+     *
+     * - length of gcoap_dns_ctx_t::addr_out in bytes on success
+     * - -EBADMSG, when receiving erroneous response or response containing
+     * - -EDESTADDRREQ, if CoAP response was received from an unexpected remote.
+     * - -EINVAL, when block-wise transfer can not be completed.
+     * - -ENOBUFS, if length of received CoAP body is greater than
+     *   @ref CONFIG_DNS_MSG_LEN.
+     * - -ENOMSG, if CoAP response did not contain a DNS response.
+     * - -ETIMEDOUT, if CoAP request timed out.
+     */
+    int res;
+    uint8_t method;         /**< CoAP method to be used for the DNS message
+                             *   exchange */
+    uint8_t dns_buf_len;    /**< Length of gcoap_dns_ctx_t::dns_buf */
+    int8_t family;          /**< Address family to resolve */
+    /**
+     * @brief The current block number for block-wise transfer
+     *
+     * Leave unset on function call.
+     */
+    uint8_t cur_blk_num;
+};
+
+/**
+ * @brief   Query a domain name via CoAPS synchronously
+ *
+ * @note    Only available with module `gcoap_dns_async` compiled in.
+ *
+ * @param[in] domain_name   A '\0'-terminated domain name. Must not be NULL.
+ * @param[out] addr_out     The resolved address. Must not be NULL.
+ * @param[in] family        The desired address family for @p addr_out.
+ *                          @ref AF_UNSPEC for any address family (an IPv6
+ *                          address will take preference over an IPv4 address).
+ * @param[in] method        CoAP method to use. Allowed values are
```

Comment:
To make GET and POST (which are already being removed from the draft) easy to remove without breaking other users, I suggest removing the parameter, having the actual implementation in a `_gcoap_dns_query_anymeth(..., method)` implementation and offering `gcoap_dns_query_{get,post}` as easily deprecatable (or even already deprecated) drop-in alternatives to `gcoap_dns_query` (which can then have very short documentations a la "@brief Query a dmina name via CoAPS using the POST method / Apart from the method used, this is identical to @ref gcoap_dns_query, including parameters and return values.")